pkg/planner/core/find_best_task.go

```diff
@@ -835,6 +836,172 @@ func (ds *DataSource) isMatchProp(path *util.AccessPath, prop *property.Physical
+	//	2: If the prop is empty (means no sort requirement), we will generate a random index partial combination
+	//	path from all alternatives in case that no index merge path comes out.
+
+	// Execution part doesn't support the merge operation for intersection case yet.
+	if path.IndexMergeIsIntersection {
+		return nil, false
+	}
+
+	noSortItem := prop.IsSortItemEmpty()
+	allSame, _ := prop.AllSameOrder()
+	if !allSame {
+		return nil, false
+	}
+	// step1: match the property from all the index partial alternative paths.
+	determinedIndexPartialPaths := make([]*util.AccessPath, 0, len(path.PartialAlternativeIndexPaths))
+	usedIndexMap := make(map[int64]struct{}, 1)
+	type idxWrapper struct {
+		matchIdx []int
+		pathIdx  int
+	}
+	allMatchIdxes := make([]idxWrapper, 0, len(path.PartialAlternativeIndexPaths))
+	// special logic for alternative paths:
+	// index merge:
+	//  path1: {pk, index1}
+	//  path2: {pk}
+	// if we choose pk in the first path, then path2 has no choice but pk, this will result in all single index failure.
+	// so we should collect all match prop paths down, stored as matchIdxes here.
+	for pathIdx, oneItemAlternatives := range path.PartialAlternativeIndexPaths {
+		matchIdxes := make([]int, 0, 1)
+		for i, oneIndexAlternativePath := range oneItemAlternatives {
+			// if there is some sort items and this path doesn't match this prop, continue.
+			if !noSortItem && !ds.isMatchProp(oneIndexAlternativePath, prop) {
+				continue
+			}
+			// two possibility here:
+			// 1. no sort items requirement.
+			// 2. matched with sorted items.
+			matchIdxes = append(matchIdxes, i)
+		}
+		if len(matchIdxes) == 0 {
+			// if all index alternative of one of the cnf item's couldn't match the sort property,
+			// the entire index merge union path can be ignored for this sort property, return false.
+			return nil, false
+		}
+		if len(matchIdxes) > 1 {
+			// if matchIdxes greater than 1, we should sort this match alternative path by its CountAfterAccess.
+			tmpOneItemAlternatives := oneItemAlternatives
+			slices.SortFunc(matchIdxes, func(a, b int) int {
+				lhsCountAfter := tmpOneItemAlternatives[a].CountAfterAccess
+				if len(tmpOneItemAlternatives[a].IndexFilters) > 0 {
+					lhsCountAfter = tmpOneItemAlternatives[a].CountAfterIndex
+				}
+				rhsCountAfter := tmpOneItemAlternatives[b].CountAfterAccess
+				if len(tmpOneItemAlternatives[b].IndexFilters) > 0 {
+					rhsCountAfter = tmpOneItemAlternatives[b].CountAfterIndex
+				}
+				return cmp.Compare(lhsCountAfter, rhsCountAfter)
+			})
+		}
+		allMatchIdxes = append(allMatchIdxes, idxWrapper{matchIdxes, pathIdx})
+	}
+	// sort allMatchIdxes by its element length.
+	// index merge:                index merge:
+	//  path1: {pk, index1}  ==>    path2: {pk}
+	//  path2: {pk}                 path1: {pk, index1}
+	// here for the fixed choice pk of path2, let it be the first one to choose, left choice of index1 to path1.
```

Comment:
I'm a bit skeptical about whether we really need to deliberately avoid choosing the same index.

Reply:
i think if we have a>2 or a<1, we can directly derive a $(-\infty, 1) \cup (2, +\infty)$ on a simple index range scan of a.
rather than index merge { $a1:  (-\infty, 1),  a2: (2, +\infty)$ }

old logic, just keep them identical here.

Reply:
That makes sense. However, if so, do we really need this sorting logic? I can't come up with any real cases that need this logic.

Reply:
sorting according to valid candidate length here can avoid the miss-pick across paths.
 
in the deriveStats, we already checked the simple index rule before, when generating a concrete path here, we should successfully generate it out, rather than “can't find a proper physical plan” caused by miss-pick here, saying 	
```
// index merge:                index merge:
//  path1: {pk, index1}  ==>    path2: {pk}
//  path2: {pk}                 path1: {pk, index1}
```
this index merge was validated before, when generating concrete paths, we should choose a path for path2 first, otherwise when picking pk for path1 is done, path2 has no choices then. (error)

Reply:
I mean if there are real examples like SQLs?
In such a case, I think, can we just choose a single-index plan on `pk`, or just allow using two partial paths using the same index?

Reply:
> I mean if there are real examples like SQLs? 
the modification here comes from an old unit test case failure, so I changed the logic like this.

the main reason is that: the user specified an index merge hint, then we pruned other paths in deriveStats there, the only left path is validated for index merge choosing (not all a simple index, otherwise, index merge hint is inapplicable right?), so the concrete path should successfully be chosen by a kind of order that we want before.